### PR TITLE
RSPEC-1751 detect loops with at most one iteration

### LINKED ISSUE #62: Rule S1751: Loops with at most one iteration should be refactored
Implements [RSPEC-1751](https://jira.sonarsource.com/browse/RSPEC-1751)

### PR DESCRIPTION
**Pull request by ghost**
Fixes #62

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java [hunk 1]
```diff
@@ -18,47 +18,55 @@ public UnconditionalJumpStatementCheck() {
...
+    return statement.is(UastNode.Kind.IF)
+      && statement.children.stream()
+        .filter(n -> n.is(UastNode.Kind.BLOCK, UastNode.Kind.STATEMENT))
+        .findFirst()
+        .flatMap(n -> n.children.stream().filter(UnconditionalJumpStatementCheck::isContinueOrIfHavingContinue).findFirst())
+        .isPresent();
+  }
+
+  private static boolean isContinueOrIfHavingContinue(UastNode statement) {
+    return statement.is(UastNode.Kind.CONTINUE) || statement.is(UastNode.Kind.IF)
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java]: Code Smell Code Smell: Add parentheses to make the operator precedence explicit. ([squid:S864](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=173&issues=AWJnUcQDRF79YA3jhpEQ&open=AWJnUcQDRF79YA3jhpEQ)

[ISSUE_KEY]: AWJnUcQDRF79YA3jhpEQ

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java [hunk 2]
```diff
@@ -18,47 +18,55 @@ public UnconditionalJumpStatementCheck() {
...
+      && statement.children.stream()
+        .filter(n -> n.is(UastNode.Kind.BLOCK, UastNode.Kind.STATEMENT))
+        .findFirst()
+        .flatMap(n -> n.children.stream().filter(UastNode.Kind.CONTINUE).findFirst())
+        .isPresent();
+  }
+
+  private boolean isLoopWithContinue(UastNode statement) {
+    // TODO implement what the method name suggests to prevent many false negatives
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=173&issues=AWJnUcQDRF79YA3jhpER&open=AWJnUcQDRF79YA3jhpER)

[ISSUE_KEY]: AWJnUcQDRF79YA3jhpER

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java [hunk 3]
```diff
@@ -18,47 +18,55 @@ public UnconditionalJumpStatementCheck() {
...
+  private static boolean isContinueOrIfHavingContinue(UastNode statement) {
+    return statement.is(UastNode.Kind.CONTINUE) || statement.is(UastNode.Kind.IF)
+      && statement.children.stream()
+        .filter(n -> n.is(UastNode.Kind.BLOCK, UastNode.Kind.STATEMENT))
+        .findFirst()
+        .flatMap(n -> n.children.stream().filter(UastNode.Kind.CONTINUE).findFirst())
+        .isPresent();
+  }
+
+  private boolean isLoopWithContinue(UastNode statement) {
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java]: Code Smell Code Smell: Make "isLoopWithContinue" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=173&issues=AWJnUcQDRF79YA3jhpES&open=AWJnUcQDRF79YA3jhpES)

[ISSUE_KEY]: AWJnUcQDRF79YA3jhpES

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java [hunk 4]
```diff
@@ -18,47 +18,56 @@ public UnconditionalJumpStatementCheck() {
...
+      || (statement.is(UastNode.Kind.IF)
+        && statement.children.stream()
+          .filter(n -> n.is(UastNode.Kind.BLOCK, UastNode.Kind.STATEMENT))
+          .findFirst()
+          .flatMap(n -> n.children.stream().filter(UastNode.Kind.CONTINUE).findFirst())
+          .isPresent());
+  }
+
+  private static boolean isLoopWithContinue(UastNode statement) {
+    // TODO implement what the method name suggests to prevent many false negatives
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=173&issues=AWJnUcQDRF79YA3jhpER&open=AWJnUcQDRF79YA3jhpER)

[ISSUE_KEY]: AWJnUcQDRF79YA3jhpER

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java [hunk 5]
```diff
@@ -18,47 +18,56 @@ public UnconditionalJumpStatementCheck() {
...
         .findFirst().filter(n -> n.is(JUMP_KINDS))
         .ifPresent(this::reportIssue);
     }
   }
 
   private void findUnconditionalJump(UastNode block) {
-    for (UastNode child : block.children) {
-      if (mayConditionallyContinue(child)) {
+    for (UastNode statement : block.children) {
+      if (isIfHavingContinue(statement) || isLoopWithContinue(statement)) {
```
Comment (saberduck) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/UnconditionalJumpStatementCheck.java]: maybe we can avoid this by just checking if there is `continue` statement in any descendant.

**Review by saberduck — COMMENTED** [review]
See my comment, otherwise LGTM

**Pull request merged**